synapse/handlers/federation.py

```diff
@@ -1004,16 +1013,47 @@ async def maybe_backfill(self, room_id, current_depth):
         sorted_extremeties_tuple = sorted(extremities.items(), key=lambda e: -int(e[1]))
         max_depth = sorted_extremeties_tuple[0][1]
 
-        # We don't want to specify too many extremities as it causes the backfill
-        # request URI to be too long.
-        extremities = dict(sorted_extremeties_tuple[:5])
-
-        if current_depth > max_depth:
+        # If we're approaching an extremity we trigger a backfill, otherwise we
+        # no-op.
+        if current_depth - 2 * limit > max_depth:
             logger.debug(
-                "Not backfilling as we don't need to. %d < %d", max_depth, current_depth
+                "Not backfilling as we don't need to. %d < %d - 2 * %d",
+                max_depth,
+                current_depth,
+                limit,
             )
             return
 
+        logger.debug(
+            "room_id: %s, backfill: current_depth: %s, max_depth: %s, extrems: %s",
+            room_id,
+            current_depth,
+            max_depth,
+            sorted_extremeties_tuple,
+        )
+
+        # We ignore extremities that have a greater depth than our current depth
+        # as:
```

Comment:
What's the user-facing result of this? Would there be events missing from the backfill?

Reply:
So the idea is that backfilling events from an extremity with a greater depth will likely only return events with a greater depth, and so those events wouldn't be included in the pagination request that triggered the backfill. We also assume we've already tried to backfill from those extremities due to previous pagination requests, and therefore failed to get anywhere.

If for some reason those things aren't true then yes, there'll be a gap in the graph that we don't tell the clients about. Note: this can happen anyway